### PR TITLE
Making client generic over database service. Part I - traitify & close

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -77,11 +78,17 @@ impl ClientReport {
 	}
 }
 
+#[cfg(not(feature = "ipc-disable"))]
```
Comment (NikVolf) [inline ethcore/src/client/client.rs]: key is to close client on specific database service here

Comment (NikVolf) [inline ethcore/src/client/client.rs]: key is to close client on specific database service here

File: ethcore/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -507,7 +506,7 @@ impl BlockChain {
 	}
 
 	/// Iterator that lists `first` and then all of `first`'s ancestors, by hash.
-	pub fn ancestry_iter(&self, first: H256) -> Option<AncestryIter> {
+	pub fn ancestry_iter<'a>(&'a self, first: H256) -> Option<AncestryIter<'a, D>> {
```
Comment (rphmeier) [inline ethcore/src/blockchain/blockchain.rs]: I think this lifetime can be elided

File: ethcore/src/client/client.rs [hunk 3]
```diff
@@ -124,12 +133,18 @@ pub fn append_path(path: &Path, item: &str) -> String {
 	p.to_str().unwrap().to_owned()
 }
 
-impl<V> Client<V> where V: Verifier {
+impl<D: Deref + Send + Sync, V: Verifier> Client<D, V> where D::Target: DatabaseService + Sized {
```
Comment (rphmeier) [inline ethcore/src/client/client.rs]: consider requiring a Sized bound on DatabaseService to save some typing

**Comment by NikVolf** [comment]
@rphmeier 
cool notes, thanx
resolved

**Pull request closed without merging**